### PR TITLE
resource_control: introduce privilege RESOURCE_GROUP_USER to restrict the switch of resource group

### LINKED ISSUE #53440: Permission enforcement of switching resource group
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

TiDB should restrict the usage of `SET RESOURCE GROUP` and `/*+RESOURCE_GROUP(rg) */` to avoid misusage. Use should have certain privilege in order to change the resource group of current session.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?

Issue Number: close #53440

Problem Summary:

### What changed and how does it work?
This PR introduces a new dynamic privilege `RESOURCE_GROUP_USER` to restrict switch current session or query's resource group.
- Only user with `RESOURCE_GROUP_USER` or `RESOURCE_GROUP_ADMIN` can execute `SET RESOURCE GROUP name` statement.
- Only user with `RESOURCE_GROUP_USER` or `RESOURCE_GROUP_ADMIN` can use  optimizer hint `RESOURCE_GROUP(name)`, this hint will be ignored and return a warning if the user doesn't has this two privilege.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Introduce privilege `RESOURCE_GROUP_USER` to restrict the switch of resource group
```

### CODE REVIEW

**Comment by glorv** [comment]
@tiancaiamao @nolouch  PTAL

File: pkg/planner/core/planbuilder.go [hunk 1]
```diff
@@ -78,10 +79,21 @@ type visitInfo struct {
...
 }
 
+func (v *visitInfo) Equals(other *visitInfo) bool {
+	return v.privilege == other.privilege &&
+		v.db == other.db &&
+		v.table == other.table &&
+		v.column == other.column &&
+		v.err == other.err &&
+		v.alterWritable == other.alterWritable &&
+		reflect.DeepEqual(v.dynamicPrivs, other.dynamicPrivs) &&
```
Comment (nolouch) [inline pkg/planner/core/planbuilder.go:93]: If this function is only for testing, DeepEqual can be directly used in the test.

File: pkg/executor/simple.go [hunk 2]
```diff
@@ -1686,8 +1686,8 @@ func (e *SimpleExec) executeAlterUser(ctx context.Context, s *ast.AlterUserStmt)
 	}
 	activeRoles := e.Ctx().GetSessionVars().ActiveRoles
 	hasCreateUserPriv := checker.RequestVerification(activeRoles, "", "", "", mysql.CreateUserPriv)
-	hasSystemUserPriv := checker.RequestDynamicVerification(activeRoles, "SYSTEM_USER", false)
-	hasRestrictedUserPriv := checker.RequestDynamicVerification(activeRoles, "RESTRICTED_USER_ADMIN", false)
+	hasSystemUserPriv := checker.RequestDynamicVerification(activeRoles, []string{"SYSTEM_USER"}, false)
+	hasRestrictedUserPriv := checker.RequestDynamicVerification(activeRoles, []string{"RESTRICTED_USER_ADMIN"}, false)
```
Comment (nolouch) [inline pkg/executor/simple.go:1690]: It seems that we can also use the old interface to check a list of privileges? like here

**Review by nolouch — COMMENTED** [review]
RESOURCE_GROUP_ADMIN:  CREATE/DROP/ALTER/SET/HINT resource group, CALIBRATE resources.

 RESOURCE_GROUP_USER: SET/HINT resource group

am I right?

File: pkg/planner/core/planbuilder.go (hunk 1 above)
Comment (glorv) [reply pkg/planner/core/planbuilder.go:93]: Can't directly use DeepEqual because `v.err` is `error` type and should not use DeepEqual for error.

File: pkg/executor/simple.go (hunk 2 above)
Comment (glorv) [reply pkg/executor/simple.go:1690]: Yes, it doable, but not very convenient in some cases. While except for ResourceGroup there is only 1 other cases that need to pass 2 permissions. I'm also not sure if there are other cases in the future.

**Comment by glorv** [comment]
> RESOURCE_GROUP_ADMIN: CREATE/DROP/ALTER/SET/HINT resource group, CALIBRATE resources.
> 
> RESOURCE_GROUP_USER: SET/HINT resource group
> 
> am I right?

Yes

File: pkg/executor/simple.go (hunk 2 above)
Comment (nolouch) [reply pkg/executor/simple.go:1690]: cc @tiancaiamao Do you think this modification is okay?

**Comment by glorv** [comment]
@tiancaiamao  PTAL, thanks~

File: pkg/executor/simple.go (hunk 2 above)
Comment (tiancaiamao) [reply pkg/executor/simple.go:1690]: Maybe change the interface is unnecessary.
We do not have much cases relying on the new interface.

File: pkg/planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -7288,10 +7288,10 @@ func collectTableName(node ast.ResultSetNode, updatableName *map[string]bool, in
 	}
 }
 
-func appendDynamicVisitInfo(vi []visitInfo, priv string, withGrant bool, err error) []visitInfo {
+func appendDynamicVisitInfo(vi []visitInfo, privs []string, withGrant bool, err error) []visitInfo {
 	return append(vi, visitInfo{
 		privilege:        mysql.ExtendedPriv,
-		dynamicPriv:      priv,
+		dynamicPrivs:     privs,
```
Comment (tiancaiamao) [inline pkg/planner/core/logical_plan_builder.go:7294]: Is this change necessary? Maybe you can 'git checkout master xxx.go' for most of the changes. @glorv

Comment (glorv) [reply pkg/planner/core/logical_plan_builder.go:7294]: Yes, I think this change is necessary to express that user should have at least one of the privileges. I added a comment for the field `dynamicPrivs`

**Review by tiancaiamao — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]
